sklearn/metrics/pairwise.py

```diff
@@ -256,6 +277,150 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
     return distances if squared else np.sqrt(distances, out=distances)
 
 
+# Pairwise distances in the presence of missing values
+def masked_euclidean_distances(X, Y=None, squared=False,
+                               missing_values="NaN", copy=True):
+    """Calculates euclidean distances in the presence of missing values
+
+    Considering the rows of X (and Y=X) as vectors, compute the
+    distance matrix between each pair of vectors. Similarly, if
+    Y is not X, then compute the distance matrix between each
+    pair of vectors (i.e., each row pair) in X and Y.
+
+    This function computes pairwise euclidean distance for vectors
+    in dense matrices X and Y with missing values in arbitrary
+    coordinates.
+
+    The following formula is used for this:
+
+        dist(X, Y) = (X.shape[1] / ((dot(NX, NYT)))) *
+                    (dot((X * X), NYT) - 2 * (dot(X, Y.T)) +
+                    dot(NX, (Y.T * Y.T)))
+
+    where NX and NYT represent the logical-not of the missing masks of
+    X and Y.T, respectively.
+    Formula in matrix form derived by:
+    Shreya Bhattarai <shreya.bhattarai@gmail.com>
+
+    This formulation zero-weights feature coordinates with missing value in
+    either vector in the pair and up-weights the remaining coordinates.
+    For instance, say we have two sample points (x1, y1) and (x2, NaN):
+
+    To calculate the euclidean distance between these, first the square
+    "distance" is calculated based only on the first feature coordinate
+    as the second coordinate is missing in one of the samples,
+    i.e., we have (x2-x1)**2. This squared distance is scaled-up by the ratio
+    of total number of coordinates to the number of available coordinates,
+    which in this case is 2/1 = 2. Now, we are left with 2*((x2-x1)**2).
+    Finally, if squared=False then the square root of this is evaluated
+    and returned otherwise the value is returned as is.
+
+    Breakdown of euclidean distance calculation between a vector pair x,y:
+
+        weight = Total # of coordinates / # of non-missing coordinates
+        dist(x,y) = sqrt(weight * sq. distance from non-missing coordinates)
+
+    This formulation implies that if all coordinates are missing in either
+    vector in the pair or if there are no common non-missing coordinates then
+    NaN is returned for that pair.
+
+    References
```

Comment:
We'd usually put this after Returns, before See Also (see for example additive_chi2_kernel in this file).